features/support/env.rb

```diff
@@ -42,6 +42,15 @@
   log.error(message + "failed")  if scenario.failed?
 end
 
+After('@users') do
+  user_name = config["features"]["users"]["user_name"]
+  users = control_node.openstack.user.list
+  user_names = users.map(&:name)
+  if user_names.include? user_name
+    control_node.openstack.user.delete user_name
+  end
+end
+
```

Comment:
This is exactly the same discussion we already had at length in #2, except this time it's about users instead of images.

> But wouldn't that mean that the cleaning-up part is not executed after all tests are finished. So test_user would still exist then and only get deleted at a rerun?

Yes, correct.  I don't see a problem with that approach in general; in fact it's quite a nice trick and I'm glad @vmoravec discovered it in the upstream Cucumber community.  For images you might risk exceeding quotas, but again the quota usage can be checked in the background.

As a general principle it seems that unlike @vmoravec, I think it can be fine to use tagged hooks.  However in this case I think I agree with him, because the tag is too broad to have a user deletion hook attached to it.  A more suitable tag name would be `@cleanup_users` or similar.  A `@users` tag should be used to tag *all* scenarios which involve testing user management of any sort, and not all of those scenarios would create a new user and consequently require user deletion.